lib/ipmi_backend_utils.pm

```diff
@@ -70,79 +68,86 @@ sub setup_console_in_grub {
     my $grub_default_file = "${root_dir}/etc/default/grub";
     my $grub_cfg_file = "";
     my $com_settings = "";
-    my $bootmethod = "";
-    my $search_pattern = "";
     my $cmd = "";
-    if ($grub_ver eq "grub2") {
-        #grub2
-        $grub_cfg_file = "${root_dir}/boot/grub2/grub.cfg";
-        if (${virt_type} eq "xen") {
-            # On some special beremetal machines, such as unreal2/3, their serial console:
-            # SERIALDEV='ttyS2', XEN_SERIAL_CONSOLE="com1=115200,8n1,0x3e8,5 console=com1"
-            $com_settings = get_var("XEN_SERIAL_CONSOLE", "console=com2,115200");
-            $bootmethod = "module";
-            $search_pattern = "vmlinuz";
+    $grub_cfg_file = "${root_dir}/boot/grub2/grub.cfg";
+    script_run("cp $grub_cfg_file ${grub_cfg_file}.org");
+    if (${virt_type} eq "xen") {
+
+        # Setting grub menuentry selection on sol console with grub2-set-default as xen, during host installation
+        if (is_xen_host && get_var('XEN_DEFAULT_BOOT_IS_SET')) {
+            $cmd = "sed -i '/### END \\\/etc\\\/grub.d\\\/00_header ###/iset default=2' $grub_cfg_file";
+            assert_script_run($cmd);
+        }
+
+        # Set serial console in kernel options
+        # On some special beremetal machines, such as unreal2/3, their serial console:
+        # SERIALDEV='ttyS2', XEN_SERIAL_CONSOLE="com1=115200,8n1,0x3e8,5 console=com1"
+        $com_settings = get_var("XEN_SERIAL_CONSOLE", "console=com2,115200");
 
+        if (is_uefi_boot) {
+            # Xen host with UEFI boot uses different grub2 configurate file on a seperate partion
+            if (${root_dir} eq "/") {
+                ${root_dir} = "/boot/efi";
+            }
+            else {
+                ${root_dir} = "/efi_mnt";
+                script_run("mkdir -p ${root_dir}");
+                my $y2log_file = '/var/log/YaST2/y2log';
+                $cmd = qq{grep -o '/dev/[^ ]\\+ /mnt/boot/efi ' $y2log_file | head -n1 | cut -f1 -d' '};
+                my $efi_partition = script_output("$cmd");
+                script_run("mount $efi_partition ${root_dir}");
+            }
+            script_run("find ${root_dir}");
+            my $xen_efi_grub_cfg_file = script_output("ls ${root_dir}/efi/sles/xen-*.cfg");
+            $cmd
+              = "sed -ri '/options=/ "
+              . "{s/(com[0-2]|console|loglevel|loglvl|guest_loglvl)=[^ ]* //g; "
+              . "/options=/ s/\$/ $com_settings loglvl=all guest_loglvl=all sync_console/;}; "
+              . "' $xen_efi_grub_cfg_file";
+            assert_script_run($cmd);
+            script_run("cat $xen_efi_grub_cfg_file");
+            upload_logs($xen_efi_grub_cfg_file);
```

Comment:
Code LGTM. 
Only one question about when `setup_console_in_grub` is called. From a xen verification job, I see xen-4.18.2_02-150600.1.cfg only upload once and  uploaded in update_package module. See [update_package-xen-4.18.2_02-150600.1.cfg](https://10.145.10.207/tests/14676208/file/update_package-xen-4.18.2_02-150600.1.cfg). Does this mean that we do not have xen settings adjusted earlier for serial console , eg during host installation and before login_console. If this is the case, will the xen serial console log collecting be good before update_package?

Reply:
the case you mentioned is an autoyast installation, so `setup_console_in_grub` was not called as expected. You can see this case, https://10.145.10.207/tests/14704604#downloads, xen-*.cfg was uploaded both in host installation and update_package.

Reply:
Our main tests rely on autoyast installation, right? Can it be ensured there too?

Reply:
Our tests support both autoyast and non-autoyast installations, and autoyast-installation is its default. I checked, if we set the correct bootloader type, legace BIOS boot or EFI boot, the kernel parameters defined below will be writen into the grub configurartion file correctly.

```
      % if ($check_var->('SYSTEM_ROLE', 'xen')) {
      <xen_append>splash=silent quiet console=tty loglevel=5 <%= defined $bmwqemu::vars{"OPT_KERNEL_PARAMS"} ? $bmwqemu::vars{"OPT_KERNEL_PARAMS"} : "" %></xen_append>
      <!-- For some special beremetal machines, such as unreal2/3, their serial console:
      SERIALDEV='ttyS2', XEN_SERIAL_CONSOLE="com1=115200,8n1,0x3e8,5 console=com1" -->
      <xen_kernel_append><%= defined $bmwqemu::vars{"XEN_SERIAL_CONSOLE"} ? $bmwqemu::vars{"XEN_SERIAL_CONSOLE"} : "console=com2,115200" %> vga=gfx-1024x768x16 loglvl=all guest_loglvl=all sync_console</xen_kernel_append>
```

Reply:
Ok, cool.